docs/root/intro/arch_overview/matching/matching_api.rst

```diff
@@ -0,0 +1,112 @@
+                        - predicate:
+                            or_matcher:
+                                predicate:
+                                - single_predicate:
+                                input:
+                                        name: request-headers
+                                        typed_config:
+                                            "@type": type.googleapis.com/envoy.type.matcher.v3.HttpRequestHeaderMatchInput
+                                            header_name: second-header
+                                    value_match:
+                                        exact_match: foo
+                                - single_predicate:
+                                input:
+                                        name: request-headers
+                                        typed_config:
+                                            "@type": type.googleapis.com/envoy.type.matcher.v3.HttpRequestHeaderMatchInput
+                                            header_name: second-header
+                                    value_match:
+                                        exact_match: bar
+                            action:
+                                name: skip
+                                typed_config:
+                                    "@type": type.googleapis.com/envoy.extensions.filters.common.matcher.action.v3.SkipFilter
+
+Above is a slightly more complicated example which combines a top level tree matcher with a linear matcher. While the tree matchers provide
+very efficient matching, they are not very expressive. The list matcher can be used to provide a much richer matching API, and can be combined
+with the tree matcher in an arbitrary order. The example describes the following match logic: skip the filter if `some-header: skip_filter`
+is present and `second-header` is set to *either* `foo` or `bar`.
+
+HTTP Filter Iteration Impact
+============================
+
+The above example only demonstrates matching on request headers, which ends up being the simplest case due to it happening before the associated
+filter receives any data. Matching on other HTTP input sources is supported (e.g. response headers), but some discussion is warranted on how this
+works at a filter level. 
+
+Currently the match evaluation for HTTP filters does not impact control flow at all: if insufficient data is available to perform the match,
+callbacks will be sent to the associated filter as normal. Once sufficient data is available to match an action, this is provided to the filter.
+A consequence of this is that if the filter wishes to gate some behavior on a match result, it has to manage stopping the iteration on its own.
+
+When it comes to actions such as SkipFilter, this means that if the skip condition is based on anything but the request headers, the filter might
+get partially applied until the match result is ready. This might result in surprising beahvior.
```

Comment:
Surprises are fun :) Is there a way to call out some concrete examples of where this happens?

Reply:
Sure I'll dig through and find an explicit example. 

Down the line I would like to use https://github.com/envoyproxy/envoy/issues/14727 to make it possible for filters to codify these to cause config rejection instead of these surprises, but this is what we've got for now